src/numbuf/python/src/pynumbuf/numbuf.cc

```diff
@@ -31,30 +33,30 @@ PyObject* NumbufPlasmaObjectExistsError;
 using namespace arrow;
 using namespace numbuf;
 
-int64_t make_schema_and_batch(std::shared_ptr<Array> data,
-    std::shared_ptr<Buffer>* metadata_out, std::shared_ptr<RecordBatch>* batch_out) {
+std::shared_ptr<RecordBatch> make_batch(std::shared_ptr<Array> data) {
   auto field = std::make_shared<Field>("list", data->type());
   std::shared_ptr<Schema> schema(new Schema({field}));
-  *batch_out =
-      std::shared_ptr<RecordBatch>(new RecordBatch(schema, data->length(), {data}));
-  int64_t size = 0;
-  ARROW_CHECK_OK(ipc::GetRecordBatchSize(batch_out->get(), &size));
-  ARROW_CHECK_OK(ipc::WriteSchema((*batch_out)->schema().get(), metadata_out));
+  return std::shared_ptr<RecordBatch>(new RecordBatch(schema, data->length(), {data}));
+}
+
+int64_t get_batch_size(std::shared_ptr<RecordBatch> batch) {
+  // Determine the size of the file by writing to a mock file.
+  auto mock = std::make_shared<MockBufferStream>();
+  std::shared_ptr<arrow::ipc::FileWriter> writer;
+  ipc::FileWriter::Open(mock.get(), batch->schema(), &writer);
+  writer->WriteRecordBatch(*batch);
+  writer->Close();
+  int64_t size;
+  ARROW_CHECK_OK(mock->Tell(&size));
   return size;
 }
 
-Status read_batch(std::shared_ptr<Buffer> schema_buffer, int64_t header_end_offset,
-    uint8_t* data, int64_t size, std::shared_ptr<RecordBatch>* batch_out) {
-  std::shared_ptr<ipc::Message> message;
-  RETURN_NOT_OK(ipc::Message::Open(schema_buffer, &message));
-  DCHECK_EQ(ipc::Message::SCHEMA, message->type());
-  std::shared_ptr<ipc::SchemaMessage> schema_msg = message->GetSchema();
-  std::shared_ptr<Schema> schema;
-  RETURN_NOT_OK(schema_msg->GetSchema(&schema));
-  auto source = std::make_shared<FixedBufferStream>(data, size);
-  std::shared_ptr<arrow::ipc::RecordBatchReader> reader;
-  RETURN_NOT_OK(ipc::RecordBatchReader::Open(source.get(), header_end_offset, &reader));
-  RETURN_NOT_OK(reader->GetRecordBatch(schema, batch_out));
+Status read_batch(uint8_t* data, int64_t size, std::shared_ptr<RecordBatch>* batch_out) {
+  std::shared_ptr<arrow::ipc::FileReader> reader;
+  auto source = std::make_shared<FixedBufferStream>(sizeof(size) + data, size - sizeof(size));
+  int64_t data_size = *((int64_t*) data);
+  arrow::ipc::FileReader::Open(source, data_size, &reader);
+  reader->GetRecordBatch(0, batch_out);
```

Comment:
The API will be something else once I finish ARROW-661, I will post guidance on what to change this to once that is done